### ISSUE TITLE
Call to overloaded function with a union is invalid

### ISSUE THREAD

**Issue by refi64** [microsoft/pyright#1588]
**Describe the bug**

If I have some overloaded function `f` that takes one argument, overloaded to be either `A` or `B`, and I call it with `Union[A, B]`, the call is marked as invalid.

**To Reproduce**

Created the following files:

`typings/tstmod/__init__.pyi`:

```python
from typing import overload

@overload
def f(x: str) -> str: pass

@overload
def f(x: bytes) -> str: pass
```

`tst.py`:

```python
# pyright: strict
from typing import Union
import tstmod

def f(x: Union[str, bytes]) -> str:
    return tstmod.f(x)
```

This results in:

```
  /var/home/ryan/playground/pyright/tst.py:6:12 - error: No overloads for "f" match the provided arguments
    Argument types: (str | bytes) (reportGeneralTypeIssues)
  /var/home/ryan/playground/pyright/tst.py:6:12 - error: Return type is unknown (reportUnknownVariableType)
```

**Expected behavior**

Code should type check.

**VS Code extension or command-line**

This has been tested with pyright 1.118 and pyright 1.116 included with PyLance.

**Additional context**

This has been tested to work on mypy 0.812.

**Comment by erictraut** [comment]
This behavior is intended. Please refer to https://github.com/microsoft/pyright/issues/1565.

**Comment by refi64** [comment]
Ah okay, that's a bit of a bummer. This is coming up for me with `urllib.parse.quote()`, which can take either `str` or `bytes` but fails to type check when given a `Union[str, bytes]`.

**Comment by erictraut** [comment]
After receiving a number of bug reports on this issue, I've decided to change the behavior of overload matching to more closely mirror that of other type checkers. New versions of of pyright will perform union expansion and evaluate multiple overloads. I've addressed the combinatoric explosion problem by limiting expansion to 64 combinations. In practice, this will occur very infrequently because it would require more than 6 arguments with two union subtypes each or more than 2 arguments with 8 union subtypes each.

**Comment by erictraut** [comment]
This change is included in pyright 1.1.135, which I just published. It will also be included in the next release of pylance.
